### PR TITLE
Connection testing harness

### PR DESCRIPTION
**Pull request by JonathanHenson**
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: tests/connection_state_test.c [hunk 1]
```diff
@@ -0,0 +1,947 @@
...
+    aws_mutex_unlock(&state_test_data->lock);
+    aws_condition_variable_notify_one(&state_test_data->cvar);
+}
+
+static bool s_is_connection_interupted(void *arg) {
+    struct mqtt_connection_state_test *state_test_data = arg;
+    return state_test_data->connection_interupted;
+}
+
+static void s_wait_for_interupt_to_complete(struct mqtt_connection_state_test *state_test_data) {
```
Comment (justinboswell) [inline tests/connection_state_test.c]: interrupt

File: tests/connection_state_test.c [hunk 2]
```diff
@@ -0,0 +1,947 @@
...
+    struct aws_socket_endpoint endpoint;
+    struct aws_socket *listener;
+    struct aws_mqtt_client mqtt_client;
+    struct aws_mqtt_client_connection *mqtt_connection;
+    struct aws_socket_options socket_options;
+    bool session_present;
+    bool connection_completed;
+    bool client_disconnect_completed;
+    bool server_disconnect_completed;
+    bool connection_interupted;
```
Comment (justinboswell) [inline tests/connection_state_test.c]: interrupted

Also, interrupt/resume can happen multiple times, should these be counters? For that matter, should connection_completed, subscribe_completed, and so on also be counters?

File: tests/connection_state_test.c [hunk 3]
```diff
@@ -0,0 +1,947 @@
...
+    bool connection_resumed;
+    bool subscribe_completed;
+    int interuption_error;
```
Comment (justinboswell) [inline tests/connection_state_test.c]: interruption_error

File: tests/connection_state_test.c [hunk 4]
```diff
@@ -0,0 +1,947 @@
...
+static void s_wait_for_publish(struct mqtt_connection_state_test *state_test_data) {
+    aws_mutex_lock(&state_test_data->lock);
+    aws_condition_variable_wait_pred(
+        &state_test_data->cvar, &state_test_data->lock, s_is_publish_received, state_test_data);
+    state_test_data->publishes_received = 0;
+    state_test_data->expected_publishes = 0;
+    aws_mutex_unlock(&state_test_data->lock);
+}
+
+static void s_on_suback(
```
Comment (justinboswell) [inline tests/connection_state_test.c]: Should we also track unsubacks? Probably right?

File: tests/connection_state_test.c (hunk 2 above)
Comment (JonathanHenson) [reply tests/connection_state_test.c]: they get flipped after they're used. They're only used for semaphores so counters wouldn't buy us anything.

File: tests/connection_state_test.c (hunk 4 above)
Comment (JonathanHenson) [reply tests/connection_state_test.c]: we're not really testing the server.

Comment (JonathanHenson) [reply tests/connection_state_test.c]: so the act of the test not hanging tests that the client does its job.

**Pull request merged**